[CI] ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean

There are several tests from `ClassificationIT` failing (they might be correlated) - see https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+release-tests/263/ 

I've also seen quite a few failures of `ClassificationIT` over the last few months (none have been tracked though).

**Build scan**:
https://gradle-enterprise.elastic.co/s/efsbpr5sbnzwk

**Repro line**:
```./gradlew ':x-pack:plugin:ml:qa:native-multi-node-tests:integTest' --tests "org.elasticsearch.xpack.ml.integration.ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean" \
  -Dtests.seed=158A5379F8D00EFB \
  -Dtests.security.manager=true \
  -Dbuild.snapshot=false \
  -Dtests.jvm.argline="-Dbuild.snapshot=false" \
  -Dtests.locale=lt \
  -Dtests.timezone=America/Hermosillo \
  -Druntime.java=8
```

**Reproduces locally?**: No

**Applicable branches**: 7.x

**Failure history**:

**Failure excerpt**:
```
java.lang.AssertionError: Stats were: {"id":"classification_training_percent_is_50_boolean","state":"analyzing","progress":[{"phase":"reindexing","progress_percent":100},{"phase":"loading_data","progress_percent":100},{"phase":"feature_selection","progress_percent":100},{"phase":"coarse_parameter_search","progress_percent":100},{"phase":"fine_tuning_parameters","progress_percent":100},{"phase":"final_training","progress_percent":0},{"phase":"writing_results","progress_percent":0},{"phase":"inference","progress_percent":0}],"data_counts":{"training_docs_count":150,"test_docs_count":0,"skipped_docs_count":0},"memory_usage":{"timestamp":1596639663246,"peak_usage_bytes":17924,"status":"ok"},"node":{"id":"qTn19sC3Qju6ZOgwo7eDqw","name":"integTest-2","ephemeral_id":"oOFqlRycRB6L868ehr6p6w","transport_address":"127.0.0.1:34239","attributes":{"testattr":"test","ml.machine_memory":"101203308544","ml.max_open_jobs":"20","xpack.installed":"true","transform.node":"true"}},"assignment_explanation":""}
Expected: <stopped>
     but: was <analyzing>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.assertIsStopped(MlNativeDataFrameAnalyticsIntegTestCase.java:207)
	at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.lambda$waitUntilAnalyticsIsStopped$0(MlNativeDataFrameAnalyticsIntegTestCase.java:157)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:951)
	at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.waitUntilAnalyticsIsStopped(MlNativeDataFrameAnalyticsIntegTestCase.java:157)
	at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.waitUntilAnalyticsIsStopped(MlNativeDataFrameAnalyticsIntegTestCase.java:153)
	at org.elasticsearch.xpack.ml.integration.ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty(ClassificationIT.java:275)
	at org.elasticsearch.xpack.ml.integration.ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean(ClassificationIT.java:354)
```

There were a similar set of failures in https://gradle-enterprise.elastic.co/s/eikzb6eka4hk4

The `ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean` failed with this:

```
java.lang.AssertionError: Stats were: {"id":"classification_training_percent_is_50_boolean","state":"starting","progress":[{"phase":"reindexing","progress_percent":1},{"phase":"loading_data","progress_percent":0},{"phase":"feature_selection","progress_percent":0},{"phase":"coarse_parameter_search","progress_percent":0},{"phase":"fine_tuning_parameters","progress_percent":0},{"phase":"final_training","progress_percent":0},{"phase":"writing_results","progress_percent":0},{"phase":"inference","progress_percent":0}],"data_counts":{"training_docs_count":0,"test_docs_count":0,"skipped_docs_count":0},"memory_usage":{"peak_usage_bytes":0,"status":"ok"},"node":{"id":"_cUVYHvfR9WfE9JsYwk_6Q","name":"integTest-2","ephemeral_id":"T2Py-4dsRyqVIYV2kHqN9Q","transport_address":"127.0.0.1:56721","attributes":{"testattr":"test","ml.machine_memory":"101104721920","ml.max_open_jobs":"20","xpack.installed":"true","transform.node":"true"}},"assignment_explanation":""}
   Expected: <stopped>
   but: was <starting> 
```

Then 6 more tests failed afterwards with a timeout on cluster health during their test cleanup.  The reason for that is that one of the nodes dies due to an assertion error:

```
[2020-08-10T12:50:27,133][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [integTest-0] fatal error in thread [elasticsearch[integTest-0][ml_job_comms][T#1]], exiting
java.lang.AssertionError: null
        at org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference.EnsembleInferenceModel.innerInfer(EnsembleInferenceModel.java:139) ~[?:?]
        at org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference.EnsembleInferenceModel.infer(EnsembleInferenceModel.java:123) ~[?:?]
        at org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference.InferenceDefinition.infer(InferenceDefinition.java:80) ~[?:?]
        at org.elasticsearch.xpack.ml.inference.loadingservice.LocalModel.inferNoStats(LocalModel.java:121) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.inference.InferenceRunner.inferTestDocs(InferenceRunner.java:116) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.inference.InferenceRunner.run(InferenceRunner.java:88) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.process.AnalyticsProcessManager.runInference(AnalyticsProcessManager.java:343) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.process.AnalyticsProcessManager.processData(AnalyticsProcessManager.java:196) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.process.AnalyticsProcessManager.lambda$runJob$1(AnalyticsProcessManager.java:146) ~[?:?]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:651) ~[elasticsearch-7.10.0-SNAPSHOT.jar:7.10.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_241]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_241]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_241]
```

Because the tests are against a multi-node cluster the subsequent tests all succeed running against the two nodes that are left (which is good), and the only thing that fails is the check that the cluster is fully healthy.

Another failure on 7.x 

https://gradle-enterprise.elastic.co/s/aamxj7fftpyim

Similar to the analysis above subsequent tests failed because the cluster cannot get to a green state with 2 nodes 


```
java.lang.AssertionError: Stats were: {"id":"classification_training_percent_is_50_integer","state":"starting","progress":[{"phase":"reindexing","progress_percent":1},{"phase":"loading_data","progress_percent":0},{"phase":"feature_selection","progress_percent":0},{"phase":"coarse_parameter_search","progress_percent":0},{"phase":"fine_tuning_parameters","progress_percent":0},{"phase":"final_training","progress_percent":0},{"phase":"writing_results","progress_percent":0},{"phase":"inference","progress_percent":0}],"data_counts":{"training_docs_count":0,"test_docs_count":0,"skipped_docs_count":0},"memory_usage":{"peak_usage_bytes":0,"status":"ok"},"node":{"id":"7JwK-kHNQ0SRII5ZIGvn0g","name":"javaRestTest-2","ephemeral_id":"8XZaFiNoQ--_HfVw_beZGA","transport_address":"127.0.0.1:52456","attributes":{"testattr":"test","ml.machine_memory":"34359738368","ml.max_open_jobs":"20","xpack.installed":"true","transform.node":"true"}},"assignment_explanation":""}
Expected: <stopped>
     but: was <starting>
```

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/EnsembleInferenceModel.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/EnsembleInferenceModel.java
@@ -8,6 +8,7 @@ package org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference;
 
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.apache.lucene.util.RamUsageEstimator;
 import org.elasticsearch.common.Nullable;
 import org.elasticsearch.common.Strings;
@@ -132,7 +133,9 @@ public class EnsembleInferenceModel implements InferenceModel {
         if (preparedForInference == false) {
             throw ExceptionsHelper.serverError("model is not prepared for inference");
         }
-        LOGGER.debug("Inference called with feature names [{}]", Strings.arrayToCommaDelimitedString(featureNames));
+        LOGGER.debug(
+            () -> new ParameterizedMessage("Inference called with feature names [{}]", Strings.arrayToCommaDelimitedString(featureNames))
+        );
         double[][] inferenceResults = new double[this.models.size()][];
         double[][] featureInfluence = new double[features.length][];
         int i = 0;
@@ -232,27 +235,28 @@ public class EnsembleInferenceModel implements InferenceModel {
     }
 
     @Override
-    public void rewriteFeatureIndices(Map<String, Integer> newFeatureIndexMapping) {
-        LOGGER.debug("rewriting features {}", newFeatureIndexMapping);
+    public void rewriteFeatureIndices(final Map<String, Integer> newFeatureIndexMapping) {
+        LOGGER.debug(() -> new ParameterizedMessage("rewriting features {}", newFeatureIndexMapping));
         if (preparedForInference) {
             return;
         }
         preparedForInference = true;
+        Map<String, Integer> featureIndexMapping = new HashMap<>();
         if (newFeatureIndexMapping == null || newFeatureIndexMapping.isEmpty()) {
             Set<String> referencedFeatures = subModelFeatures();
-            LOGGER.debug("detected submodel feature names {}", referencedFeatures);
+            LOGGER.debug(() -> new ParameterizedMessage("detected submodel feature names {}", referencedFeatures));
             int newFeatureIndex = 0;
-            newFeatureIndexMapping = new HashMap<>();
+            featureIndexMapping = new HashMap<>();
             this.featureNames = new String[referencedFeatures.size()];
             for (String featureName : referencedFeatures) {
-                newFeatureIndexMapping.put(featureName, newFeatureIndex);
+                featureIndexMapping.put(featureName, newFeatureIndex);
                 this.featureNames[newFeatureIndex++] = featureName;
             }
         } else {
             this.featureNames = new String[0];
         }
         for (InferenceModel model : models) {
-            model.rewriteFeatureIndices(newFeatureIndexMapping);
+            model.rewriteFeatureIndices(featureIndexMapping);
         }
     }
 

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/EnsembleInferenceModel.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/EnsembleInferenceModel.java
@@ -114,19 +114,9 @@ public class EnsembleInferenceModel implements InferenceModel {
         return targetType;
     }
 
-    private double[] getFeatures(Map<String, Object> fields) {
-        double[] features = new double[featureNames.length];
-        int i = 0;
-        for (String featureName : featureNames) {
-            Double val = InferenceHelpers.toDouble(fields.get(featureName));
-            features[i++] = val == null ? Double.NaN : val;
-        }
-        return features;
-    }
-
     @Override
     public InferenceResults infer(Map<String, Object> fields, InferenceConfig config, Map<String, String> featureDecoderMap) {
-        return innerInfer(getFeatures(fields), config, featureDecoderMap);
+        return innerInfer(InferenceModel.extractFeatures(featureNames, fields), config, featureDecoderMap);
     }
 
     @Override
@@ -142,9 +132,7 @@ public class EnsembleInferenceModel implements InferenceModel {
         if (preparedForInference == false) {
             throw ExceptionsHelper.serverError("model is not prepared for inference");
         }
-        LOGGER.debug("Inference called with feature names [{}]",
-            featureNames == null ? "<null>" : Strings.arrayToCommaDelimitedString(featureNames));
-        assert featureNames != null && featureNames.length > 0;
+        LOGGER.debug("Inference called with feature names [{}]", Strings.arrayToCommaDelimitedString(featureNames));
         double[][] inferenceResults = new double[this.models.size()][];
         double[][] featureInfluence = new double[features.length][];
         int i = 0;

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/InferenceDefinition.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/InferenceDefinition.java
@@ -102,6 +102,15 @@ public class InferenceDefinition {
         }
     }
 
+    @Override
+    public String toString() {
+        return "InferenceDefinition{" +
+            "trainedModel=" + trainedModel +
+            ", preProcessors=" + preProcessors +
+            ", decoderMap=" + decoderMap +
+            '}';
+    }
+
     public static Builder builder() {
         return new Builder();
     }

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/InferenceModel.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/InferenceModel.java
@@ -10,11 +10,22 @@ import org.apache.lucene.util.Accountable;
 import org.elasticsearch.common.Nullable;
 import org.elasticsearch.xpack.core.ml.inference.results.InferenceResults;
 import org.elasticsearch.xpack.core.ml.inference.trainedmodel.InferenceConfig;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.InferenceHelpers;
 import org.elasticsearch.xpack.core.ml.inference.trainedmodel.TargetType;
 
 import java.util.Map;
 
 public interface InferenceModel extends Accountable {
+
+    static double[] extractFeatures(String[] featureNames, Map<String, Object> fields) {
+        double[] features = new double[featureNames.length];
+        int i = 0;
+        for (String featureName : featureNames) {
+            Double val = InferenceHelpers.toDouble(fields.get(featureName));
+            features[i++] = val == null ? Double.NaN : val;
+        }
+        return features;
+    }
 
     /**
      * @return The feature names in their desired order

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/TreeInferenceModel.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/TreeInferenceModel.java
@@ -127,22 +127,12 @@ public class TreeInferenceModel implements InferenceModel {
 
     @Override
     public InferenceResults infer(Map<String, Object> fields, InferenceConfig config, Map<String, String> featureDecoderMap) {
-        return innerInfer(getFeatures(fields), config, featureDecoderMap);
+        return innerInfer(InferenceModel.extractFeatures(featureNames, fields), config, featureDecoderMap);
     }
 
     @Override
     public InferenceResults infer(double[] features, InferenceConfig config) {
         return innerInfer(features, config, Collections.emptyMap());
-    }
-
-    private double[] getFeatures(Map<String, Object> fields) {
-        double[] features = new double[featureNames.length];
-        int i = 0;
-        for (String featureName : featureNames) {
-            Double val = InferenceHelpers.toDouble(fields.get(featureName));
-            features[i++] = val == null ? Double.NaN : val;
-        }
-        return features;
     }
 
     private InferenceResults innerInfer(double[] features, InferenceConfig config, Map<String, String> featureDecoderMap) {

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/TreeInferenceModel.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/inference/TreeInferenceModel.java
@@ -8,6 +8,7 @@ package org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference;
 
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.apache.lucene.util.Accountable;
 import org.elasticsearch.common.Nullable;
 import org.elasticsearch.common.Numbers;
@@ -301,7 +302,7 @@ public class TreeInferenceModel implements InferenceModel {
 
     @Override
     public void rewriteFeatureIndices(Map<String, Integer> newFeatureIndexMapping) {
-        LOGGER.debug("rewriting features {}", newFeatureIndexMapping);
+        LOGGER.debug(() -> new ParameterizedMessage("rewriting features {}", newFeatureIndexMapping));
         if (preparedForInference) {
             return;
         }
@@ -343,7 +344,7 @@ public class TreeInferenceModel implements InferenceModel {
             if (node instanceof LeafNode) {
                 LeafNode leafNode = (LeafNode) node;
                 if (leafNode.leafValue.length > 1) {
-                    return (double)leafNode.leafValue.length;
+                    return leafNode.leafValue.length;
                 } else {
                     max = Math.max(leafNode.leafValue[0], max);
                 }

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/tree/Tree.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/tree/Tree.java
@@ -165,6 +165,13 @@ public class Tree implements LenientlyParsedTrainedModel, StrictlyParsedTrainedM
             throw ExceptionsHelper.badRequestException("feature index [{}] is out of bounds for the [{}] array",
                     maxFeatureIndex, FEATURE_NAMES.getPreferredName());
         }
+        if (nodes.size() > 1) {
+            if (featureNames.isEmpty()) {
+                throw ExceptionsHelper.badRequestException("[{}] is empty and the tree has > 1 nodes; num nodes [{}]. " +
+                        "The model Must have features if tree is not a stump",
+                    FEATURE_NAMES.getPreferredName(), nodes.size());
+            }
+        }
         checkTargetType();
         detectMissingNodes();
         detectCycle();

--- a/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/tree/TreeTests.java
+++ b/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/tree/TreeTests.java
@@ -175,6 +175,35 @@ public class TreeTests extends AbstractSerializingTestCase<Tree> {
         assertThat(ex.getMessage(), equalTo(msg));
     }
 
+    public void testTreeWithEmptyFeaturesAndOneNode() {
+        // Shouldn't throw
+        Tree.builder()
+            .setRoot(TreeNode.builder(0).setLeafValue(10.0))
+            .setFeatureNames(Collections.emptyList())
+            .build()
+            .validate();
+    }
+
+    public void testTreeWithEmptyFeaturesAndThreeNodes() {
+        String msg = "[feature_names] is empty and the tree has > 1 nodes; num nodes [3]. " +
+            "The model Must have features if tree is not a stump";
+        ElasticsearchException ex = expectThrows(ElasticsearchException.class, () -> {
+            Tree.builder()
+                .setRoot(TreeNode.builder(0)
+                    .setLeftChild(1)
+                    .setRightChild(2)
+                    .setThreshold(randomDouble()))
+                .addNode(TreeNode.builder(1)
+                    .setLeafValue(randomDouble()))
+                .addNode(TreeNode.builder(2)
+                    .setLeafValue(randomDouble()))
+                .setFeatureNames(Collections.emptyList())
+                .build()
+                .validate();
+        });
+        assertThat(ex.getMessage(), equalTo(msg));
+    }
+
     public void testOperationsEstimations() {
         Tree tree = buildRandomTree(Arrays.asList("foo", "bar", "baz"), 5);
         assertThat(tree.estimatedNumOperations(), equalTo(7L));

